tests/xoreutils/test_uptime.py

```diff
@@ -0,0 +1,174 @@
+#!/usr/bin/env python
+# coding: utf8
+
+import importlib
+import sys
+import time
+from datetime import datetime
+import unittest
+
+sys.path.insert(0, ".")
+
+from xonsh.xoreutils import uptime
+
+
+boottime_helpers = [f for f in vars(uptime) if f.startswith("_boottime_")]
+uptime_helpers = [f for f in vars(uptime) if f.startswith("_uptime_")]
+
+
+class NormalTest(unittest.TestCase):
+    """	
+    This class just calls each of the functions normally and ensures they don't	
+    do dumb things like throw exceptions or return complex numbers.	
+    """
+
+    def tearDown(self):
+        """	
+        __boottime affects how boottime() and its helpers work, and it may be	
+        set as a side-effect by any function. To be on the safe side, just	
+        reload the whole module every time.	
+        """
+        importlib.reload(uptime)
+
+    def basic_test(self, func, rettypes):
+        """	
+        Calls a given function and checks if it returns something of a type	
+        in the sequence rettypes.	
+        """
+        ret = func()
+        self.assertTrue(any(isinstance(ret, t) for t in rettypes))
+
+    def __getattr__(self, name):
+        # I really don't feel like writing and maintaining over a dozen
+        # essentially identical methods, and if there's a cleaner way to do
+        # this, I couldn't find it in the unittest docs.
+        if name.startswith("test_"):
+            func = name[5:]
+            if func == "uptime" or func in uptime_helpers:
+                rettypes = (type(None), float, int)
+            elif func == "boottime" or func in boottime_helpers:
+                rettypes = (type(None), datetime)
+            else:
+                raise AttributeError()
+            return lambda: self.basic_test(getattr(uptime, func), rettypes)
+        else:
+            return unittest.TestCase.__getattr__(self, name)
+
+
+class BrokenCtypesTest(NormalTest):
+    """	
+    It's ridiculous how many platforms don't have ctypes. This class simulates	
+    that.	
+    """
+
+    @classmethod
+    def setUpClass(cls):
+        uptime.ctypes = None
+        delattr(uptime, "struct")
+        delattr(uptime, "os")
+
+
+class OtherTest(unittest.TestCase):
+    def setUp(self):
```

Comment:
pytest is preferred for the tests. You could use the fixtures xession to setup Xonsh correctly during tests.

Reply:
I´m fear about it continue to be pure lost of time for a single contributor, I have try to help and that finaly back to me like a bad joke about cross plateform ...

i contribute to other project without any trouble about cross plateform... that PR is a ultra bad experience to me. (after 25 years of Dev that the only one project with it level of complexity just for a cross plateform, Contributor spin a round without any assistance ...)

Then are you sure about convert actual tests to pytest think can help for a eventual Merge ?

If that help i´ll do ...
If that just a cosmetic fixe,  i dont touch anything .... because we are in front of a Arch/Design trouble. The code i provide can,t fixe a design trouble just by ḿigrate to pytest.

Thank for you help anyway that a plaisir to share...

Regards

If contri

Reply:
can I take over this PR? can you review perhaps. I want to merge this fix. But seems like some part of the code can be simplified.

Reply:
Please Take over that PR. That really what it PR need.

i have do my best ... but clearly need assistance
My error come from the fact that PR update a xonsh lib for Haiku-OS project in same time of my primary contribution about uname command line.

To complexe to Merge , that my bad ...

The part code you share is a copy/paste from a other test file, use as model (if i remenber well).

I confess i'm enjoy about if you can take a look on it MR

Regards
PS: nobody know but Haiku-OS user's are waiting for it MR